Gems/Atom/RHI/Code/Include/Atom/RHI/MultiDeviceQueryPool.h

```diff
@@ -0,0 +1,98 @@
+        //! When initializing multiple queries use the other version of InitQuery
+        //! because the pool will try to group the queries together.
+        //!  @param query MultiDeviceQuery to initialize.
+        ResultCode InitQuery(MultiDeviceQuery* query);
+
+        //!  Initialize a group of queries from the pool. The initialization will try to allocate
+        //!  the queries in a consecutive space (consecutive per device). The reason for this is that is more efficient when requesting
+        //!  results or copying multiple query results.
+        //!  @param queries Pointer to an array of queries to initialize.
+        //!  @param queryCount Number of queries.
+        ResultCode InitQuery(MultiDeviceQuery** queries, uint32_t queryCount);
+
+        //! Get the results from all queries (from all devices) in the pool, which are returned as uint64_t data.
+        //! The parameter "resultsCount" denotes the number of results requested per device.
+        //! The "results" parameter must contain enough space to save the results from all queries (from all devices) in the pool,
+        //! i.e. resultCount * deviceCount * sizeof(uint64_t) must be pre-allocated.
+        //! The function can return partial results. In case of failure of requesting results from a specific device, only results from
+        //! lower-indexed devices (which already have successfully returned results) are returned.
+        //! For further details related to device-specific query functionality, please check the related header.
+        ResultCode GetResults(uint64_t* results, uint32_t resultsCount, QueryResultFlagBits flags);
+
+        //! Same as GetResults(uint64_t, uint32_t, QueryResultFlagBits) but for a specific multi-device query.
+        ResultCode GetResults(MultiDeviceQuery* query, uint64_t* result, uint32_t resultsCount, QueryResultFlagBits flags);
+
+        //!  Same as GetResults(MultiDeviceQuery*, uint64_t, uint32_t, QueryResultFlagBits) but for a list of queries.
+        //!  It's more efficient if the list of queries is sorted by handle in ascending order because there's no need to sort the
+        //!  results before returning.
+        ResultCode GetResults(
+            MultiDeviceQuery** queries, uint32_t queryCount, uint64_t* results, uint32_t resultsCount, QueryResultFlagBits flags);
+
+        //!  Returns the buffer descriptor used to initialize the query pool. Descriptor contents
+        //!  are undefined for uninitialized pools.
+        const QueryPoolDescriptor& GetDescriptor() const override final;
+
+        //! Forwards the shutdown call to all device-specific QueryPools.
+        void Shutdown() override final;
+
+    private:
+        //! Validates that the queries are not null and that they belong to this pool.
+        ResultCode ValidateQueries(MultiDeviceQuery** queries, uint32_t queryCount);
+
+        QueryPoolDescriptor m_descriptor;
```

Comment:
Similar to the question with the validation, since we want to keep the functionality alive in the single-device objects, we do not want to remove it there. 
What we could do is remove the descriptor here and change the `GetDescriptor()` call such that it gets a `deviceIndex` and returns the corresponding descriptor from this device `GetDescriptor(uint32_t deviceIndex)`

Reply:
I would not do this on a `deviceIndex` basis since the descriptor should be the same for all single device objects (and thus the user should not have to care about the `deviceIndex` to get the descriptor.

We only added this member in order to be able to return the descriptor in `GetDescriptor()`. It would of course be possible to get it from one of the single device pools in the map, but we thought this is not the nicest way to do it and it would not handle the special case, where the device mask is 0 (even though that doesn't make much sense and should probably not be done).